File: lib/libsokoengine/src/game/mover.cpp
#include "mover.hpp"
#include "variant_board.hpp"
#include "board_cell.hpp"
#include "hashed_board_state.hpp"
#include "cppitertools/groupby.hpp"

#include <algorithm>

using namespace std;
using namespace iter;

namespace sokoengine {

using namespace implementation;

NonPlayableBoardError::NonPlayableBoardError():
  runtime_error("Board is not playable!")
{}

NonPlayableBoardError::~NonPlayableBoardError() = default;


IllegalMoveError::IllegalMoveError(const string& mess):
  runtime_error(mess)
{}

IllegalMoveError::~IllegalMoveError() = default;

class LIBSOKOENGINE_LOCAL Mover::PIMPL {
public:
  VariantBoard::unique_ptr_t m_initial_board;
  HashedBoardState m_state;
  SolvingMode m_solving_mode;
  bool m_pulls_boxes = true;
  piece_id_t m_selected_pusher = DEFAULT_PIECE_ID;
  size_t m_pull_count = 0;
  Mover::Moves m_last_move;

  PIMPL(VariantBoard& board, SolvingMode mode) :
    m_initial_board(board.create_clone()),
    m_state(board),
    m_solving_mode(mode)
  {
    if (!m_state.is_playable()) {
      throw NonPlayableBoardError();
    }

    if(m_solving_mode == SolvingMode::REVERSE) {
      m_state.switch_boxes_and_goals();
    }
  }

  PIMPL(const PIMPL& rv) :
    m_initial_board(rv.m_initial_board->create_clone()),
    m_state(rv.m_state),
    m_solving_mode(rv.m_solving_mode),
    m_pulls_boxes(rv.m_pulls_boxes),
    m_selected_pusher(rv.m_selected_pusher),
    m_pull_count(rv.m_pull_count),
    m_last_move(rv.m_last_move)
  {}

  PIMPL& operator=(const PIMPL& rv) {
    if (this != &rv) {
      m_initial_board = rv.m_initial_board->create_clone();
      m_state = rv.m_state;
      m_solving_mode = rv.m_solving_mode;
      m_pulls_boxes = rv.m_pulls_boxes;
      m_selected_pusher = rv.m_selected_pusher;
      m_pull_count = rv.m_pull_count;
      m_last_move = rv.m_last_move;
    }
    return *this;
  }

  PIMPL(PIMPL&& rv) = default;
  PIMPL& operator=(PIMPL&& rv) = default;

  struct MoveWorkerOptions {
    bool decrease_pull_count;
    bool increase_pull_count;
    bool force_pulls;
  };

  void select_pusher (piece_id_t pusher_id) {
    if (pusher_id == m_selected_pusher)
      return;

    position_t old_pusher_position = m_state.pusher_position(m_selected_pusher);
    position_t new_pusher_position = m_state.pusher_position(pusher_id);

    auto selection_path = m_state.board().positions_path_to_directions_path(
      m_state.board().find_jump_path(old_pusher_position, new_pusher_position)
    );

    m_last_move.clear();
    for (const Direction& direction : selection_path) {
      AtomicMove atomic_move (direction, false);
      atomic_move.set_is_pusher_selection(true);
      m_last_move.push_back(atomic_move);
    }

    m_selected_pusher = pusher_id;
  }

  void jump (position_t new_position) {
    if (m_pull_count != 0) {
      throw IllegalMoveError("Jumps not allowed after first pull");
    }

    if (m_solving_mode != SolvingMode::REVERSE) {
      throw IllegalMoveError("Jumps allowed only in reverse solving mode");
    }

    position_t old_position = m_state.pusher_position(m_selected_pusher);
    if (old_position == new_position) {
      return;
    }

    try {
      m_state.move_pusher_from(old_position, new_position);
    } catch (const CellAlreadyOccupiedError& exc) {
      throw IllegalMoveError(exc.what());
    }

    auto path = m_state.board().positions_path_to_directions_path(
      m_state.board().find_jump_path(old_position, new_position)
    );
    m_last_move.clear();
    for (const Direction& direction : path) {
      AtomicMove atomic_move(direction, false);
      atomic_move.set_is_jump(true);
      atomic_move.set_pusher_id(m_selected_pusher);
      m_last_move.push_back(atomic_move);
    }
  }

  void push_or_move(const Direction& direction, const MoveWorkerOptions& options) {
    position_t initial_pusher_position = m_state.pusher_position(
      m_selected_pusher
    );
    position_t in_front_of_pusher = m_state.board().neighbor(
      initial_pusher_position, direction
    );

    if (in_front_of_pusher == NULL_POSITION) {
      throw IllegalMoveError(
        "Can't move pusher off board! (ID: " +
        std::to_string(m_selected_pusher) + ", direction: " +
        direction.str() + ")"
      );
    }

    bool is_push = false;
    position_t in_front_of_box = NULL_POSITION;

    if (m_state.has_box_on(in_front_of_pusher)) {
      is_push = true;
      in_front_of_box = m_state.board().neighbor(
        in_front_of_pusher, direction
      );
      if (in_front_of_box == NULL_POSITION) {
        throw IllegalMoveError(
          "Can't push box off board (ID: " +
          std::to_string(m_state.box_id_on(in_front_of_pusher)) +
          ", direction: " + direction.str() + ")"
        );
      }

      try {
        m_state.move_box_from(in_front_of_pusher, in_front_of_box);
      } catch (const CellAlreadyOccupiedError& exc) {
        throw IllegalMoveError(exc.what());
      }
    }

    try {
      m_state.move_pusher_from(initial_pusher_position, in_front_of_pusher);
    } catch (const CellAlreadyOccupiedError& exc) {
      throw IllegalMoveError(exc.what());
    }

    AtomicMove atomic_move(direction, is_push);
    atomic_move.set_pusher_id(m_selected_pusher);
    if (is_push) {
      atomic_move.set_moved_box_id(m_state.box_id_on(in_front_of_box));
      if (options.decrease_pull_count && m_pull_count > 0) {
        m_pull_count -= 1;
      }
    }
    m_last_move.clear();
    m_last_move.push_back(atomic_move);
  }

  void pull_or_move(const Direction& direction, const MoveWorkerOptions& options) {
    position_t initial_pusher_position = m_state.pusher_position(
      m_selected_pusher
    );
    position_t in_front_of_pusher = m_state.board().neighbor(
      initial_pusher_position, direction
    );

    if (in_front_of_pusher == NULL_POSITION) {
      throw IllegalMoveError(
        "Can't move pusher off board! (ID: " +
        std::to_string(m_selected_pusher) + ", direction: " +
        direction.str() + ")"
      );
    }

    try {
      m_state.move_pusher_from(initial_pusher_position, in_front_of_pusher);
    } catch (const CellAlreadyOccupiedError& exc) {
      throw IllegalMoveError(exc.what());
    }

    bool is_pull = false;

    if (options.force_pulls) {
      position_t behind_pusher = m_state.board().neighbor(
        initial_pusher_position, direction.opposite()
      );

      if (behind_pusher != NULL_POSITION &&
          m_state.board().cell(behind_pusher).has_box()) {
        is_pull = true;
        try {
          m_state.move_box_from(behind_pusher, initial_pusher_position);
        } catch (const CellAlreadyOccupiedError& exc) {
          throw IllegalMoveError(exc.what());
        }
        if (options.increase_pull_count) {
          m_pull_count += 1;
        }
      }
    }

    AtomicMove atomic_move(direction, is_pull);
    atomic_move.set_pusher_id(m_selected_pusher);
    if (is_pull) {
      atomic_move.set_moved_box_id(m_state.box_id_on(initial_pusher_position));
    }
    m_last_move.clear();
    m_last_move.push_back(atomic_move);
  }

  void undo_last_move() {
    Mover::Moves new_last_moves;
    Mover::Moves old_last_moves = m_last_move;

    int jump_key = 0;
    int pusher_change_key = 1;
    int move_key = 2;

    auto key_functor = [&] (const AtomicMove& elem) {
      if (elem.is_jump()) {
        return jump_key;
      } else if (elem.is_pusher_selection()) {
        return pusher_change_key;
      }
      return move_key;
    };

    reverse(old_last_moves.begin(), old_last_moves.end());
    for (auto&& gb : groupby(old_last_moves, key_functor)) {
      // gb.first -> key
      // gb.second -> values

      if (gb.first == move_key) {
        for (const AtomicMove& am : gb.second) {
          undo_atomic_move(am);
          new_last_moves.insert(
            new_last_moves.end(), m_last_move.begin(), m_last_move.end()
          );
        }
      } else if (gb.first == jump_key) {
        Mover::Moves tmp (gb.second.begin(), gb.second.end());
        undo_jump(tmp);
        new_last_moves.insert(
          new_last_moves.end(), m_last_move.begin(), m_last_move.end()
        );
      } else {
        Mover::Moves tmp (gb.second.begin(), gb.second.end());
        undo_pusher_selection(tmp);
        new_last_moves.insert(
          new_last_moves.end(), m_last_move.begin(), m_last_move.end()
        );
      }
    }

    m_last_move = new_last_moves;
  }

  void undo_atomic_move(const AtomicMove& atomic_move) {
    MoveWorkerOptions options;
    if (m_solving_mode == SolvingMode::FORWARD) {
      bool has_box_behind_pusher = m_state.has_box_on(
        m_state.board().neighbor(
          m_state.pusher_position(m_selected_pusher),
          atomic_move.direction()
        )
      );

      if (!atomic_move.is_move() && !has_box_behind_pusher)
        throw IllegalMoveError("Requested push undo, but no box behind pusher!");

      options.force_pulls = !atomic_move.is_move();
      options.increase_pull_count = false;
      pull_or_move(atomic_move.direction().opposite(), options);
    } else {
      options.decrease_pull_count = true;
      push_or_move(atomic_move.direction().opposite(), options);
    }
  }

  void undo_jump(const Mover::Moves& jump_moves) {
    Directions path;
    for (const AtomicMove& am : jump_moves)
      path.push_back(am.direction().opposite());
    position_t old_position = m_state.pusher_position(m_selected_pusher);
    position_t new_position = m_state.board().path_destination(old_position, path);
    jump(new_position);
  }

  void undo_pusher_selection(const Mover::Moves& selection_moves) {
    Directions path;
    for (const AtomicMove& am : selection_moves)
      path.push_back(am.direction().opposite());
    position_t old_position = m_state.pusher_position(m_selected_pusher);
    position_t new_position = m_state.board().path_destination(old_position, path);
    select_pusher(m_state.pusher_id_on(new_position));
  }
};

Mover::Mover(VariantBoard& board, const SolvingMode& mode) :
  m_impl(std::make_unique<PIMPL>(board, mode))
{}

Mover::Mover(const Mover& rv) :
  m_impl(std::make_unique<PIMPL>(*rv.m_impl))
{}

Mover& Mover::operator=(const Mover& rv) {
  if (this != &rv) {
      m_impl = std::make_unique<PIMPL>(*rv.m_impl);
  }
  return *this;
}

Mover::Mover(Mover &&) = default;

Mover& Mover::operator=(Mover &&) = default;

Mover::~Mover() = default;

const VariantBoard& Mover::board() const { return m_impl->m_state.board(); }

SolvingMode Mover::solving_mode() const { return m_impl->m_solving_mode; }

const HashedBoardState& Mover::state() const { return m_impl->m_state; }

piece_id_t Mover::selected_pusher() const { return m_impl->m_selected_pusher; }

void Mover::select_pusher (piece_id_t pusher_id) {
  m_impl->select_pusher(pusher_id);
}

void Mover::jump( position_t new_position ) { m_impl->jump(new_position); }

void Mover::move( const Direction& direction ) {
  PIMPL::MoveWorkerOptions options;
  if (m_impl->m_solving_mode == SolvingMode::FORWARD) {
    options.decrease_pull_count = false;
    m_impl->push_or_move(direction, options);
  } else {
    options.force_pulls = m_impl->m_pulls_boxes;
    options.increase_pull_count = true;
    m_impl->pull_or_move(direction, options);
  }
}

void Mover::undo_last_move() { m_impl->undo_last_move(); }

const Mover::Moves& Mover::last_move() const { return m_impl->m_last_move; }

void Mover::set_last_move(const Moves& rv) { m_impl->m_last_move = rv; }

bool Mover::pulls_boxes() const { return m_impl->m_pulls_boxes; }

void Mover::set_pulls_boxes( bool value ) { m_impl->m_pulls_boxes = value; }

const VariantBoard& Mover::initial_board() const {
  return *(m_impl->m_initial_board);
}

} // namespace sokoengine

File: src/sokoenginepy/game/mover/mover.py
from enum import IntEnum
from itertools import groupby

from ...manager import DEFAULT_PIECE_ID, CellAlreadyOccupiedError, HashedBoardManager
from ...snapshot import AtomicMove


class SolvingMode(IntEnum):
    FORWARD = 0
    REVERSE = 1

    def __repr__(self):
        return "SolvingMode." + self.name


class NonPlayableBoardError(RuntimeError):
    def __init__(self):
        super().__init__("Board is not playable!")


class IllegalMoveError(RuntimeError):
    pass


class MoveWorkerOptions:
    def __init__(self):
        self.decrease_pull_count = None
        self.increase_pull_count = None
        self.force_pulls = None


class Mover:
    """Implements game rules (on-board movement).

    Supports forward and reverse solving mode.

    **Forward solving mode**

    - pusher is allowed to push single box at the time
    - pusher can't pull boxes
    - pusher can't jump over boxes or walls

    **Reverse solving mode:**

    - pusher is allowed to pull single box at the time

        - if position allows pull that pull is optional (pusher is allowed to move
          without pull even if pull is possible).
        - default behavior is to always pull boxes but that can be changed any time
          through `.pulls_boxes`

    - pusher can't push boxes

    - pusher is allowed to jump over boxes and walls

        - jumps are allowed only before first pull is done

    - board starts in solved state: positions of boxes and goals are switched

    **History management**

    Mover only stores last performed move in history and it doesn't offer redo.
    Failed moves, undo and non-moves (ie. selecting already selected pusher or
    jumping on same position pusher is already standing on) clear undo history.

    Args:
        board (VariantBoard): Instance of :class:`.VariantBoard` subclasses
        solving_mode (SolvingMode): start the game in this solving mode

    Warning:
        :class:`.Mover` operates directly on referenced game board. Because of that,
        this board should not be edited outside of :class:`.Mover` once
        :class:`.Mover` instance had been attached to it: editing the board will
        corrupt :class:`.Mover` internal state. For the same reason, it is not
        allowed to attach two movers to same game board.
    """

    def __init__(self, board, solving_mode=SolvingMode.FORWARD):
        self._manager = HashedBoardManager(board)
        self._solving_mode = solving_mode
        self._pulls_boxes = True
        self._selected_pusher = DEFAULT_PIECE_ID
        self._pull_count = 0
        self._last_move = []

        if not self._manager.is_playable:
            raise NonPlayableBoardError

        if self._solving_mode == SolvingMode.REVERSE:
            self._manager.switch_boxes_and_goals()

    @property
    def board(self):
        """Board on which :class:`.Mover` is operating on"""
        return self._manager.board

    @property
    def solving_mode(self):
        """:class:`.Mover` operation mode (:class:`.SolvingMode`)."""
        return self._solving_mode

    @property
    def board_manager(self):
        """Current board manager (:class:`.HashedBoardManager`)."""
        return self._manager

    @property
    def selected_pusher(self):
        """ID of pusher that will perform next move.

        See Also:
            :meth:`.select_pusher`
        """
        return self._selected_pusher

    @property
    def pulls_boxes(self):
        """
        Select behavior in `.SolvingMode.REVERSE` mode when pusher is moving away
        from box.

        See Also:
            :meth:`.move`
        """
        return self._pulls_boxes

    @pulls_boxes.setter
    def pulls_boxes(self, rv):
        self._pulls_boxes = rv

    @property
    def last_move(self):
        """Sequence of :class:`.AtomicMove` that contains most recent movement.

        Whenever :class:`.Mover` performs any movemet or pusher selection, it puts
        resulting :class:`.AtomicMove` into this sequence in order atomic moves
        happened.

        This is useful for movement animation in GUI. After Mover performs movement,
        GUI has enough information to know what was performed and to choose which
        animations to render for that.

        It is also possible to set this to some external sequence of moves. In that
        case, calling :meth:`.undo_last_move` will cause Mover to try to undo that
        external sequence of atomic moves.

        Example:

            >>> from sokoenginepy import Mover, SokobanBoard, AtomicMove, Direction
            >>> board = SokobanBoard(board_str='\\n'.join([
            ...     '    #####',
            ...     '    #  @#',
            ...     '    #$  #',
            ...     '  ###  $##',
            ...     '  #  $ $ #',
            ...     '### # ## #   ######',
            ...     '#   # ## #####  ..#',
            ...     '# $  $          ..#',
            ...     '##### ### #@##  ..#',
            ...     '    #     #########',
            ...     '    #######'
            ... ]))
            >>> mover = Mover(board)
            >>> mover.last_move = [AtomicMove(Direction.UP), AtomicMove(Direction.RIGHT)]
            >>> mover.undo_last_move()
            >>> mover.board
            SokobanBoard(board_str='\\n'.join([
                '    #####          ',
                '    #   #          ',
                '    #$@ #          ',
                '  ###  $##         ',
                '  #  $ $ #         ',
                '### # ## #   ######',
                '#   # ## #####  ..#',
                '# $  $          ..#',
                '##### ### #@##  ..#',
                '    #     #########',
                '    #######        '
            ]))
            >>> mover.last_move
            [AtomicMove(Direction.LEFT, box_moved=False), AtomicMove(Direction.DOWN, box_moved=False)]

        Warning:
            Subsequent movement overwrites this meaning that Mover can only undo last
            move performed (it doesn't keep whole history of movement, only the last
            move).
        """
        return self._last_move

    @last_move.setter
    def last_move(self, rv):
        self._last_move = rv

    def select_pusher(self, pusher_id):
        """
        Selects pusher that will perform next move.

        Mover always selects :data:`.DEFAULT_PIECE_ID` before any movements is
        performed. This means that for single-pusher boards, that single pusher is
        always automatically selected and this method doesn't need to be called.

        Args:
            pusher_id (int): ID of pusher

        See Also:
            `.BoardManager.pushers_ids`
        """

        if pusher_id == self._selected_pusher:
            return

        old_pusher_position = self._manager.pusher_position(self._selected_pusher)
        new_pusher_position = self._manager.pusher_position(pusher_id)
        selection_path = self._manager.board.positions_path_to_directions_path(
            self._manager.board.find_jump_path(old_pusher_position, new_pusher_position)
        )

        self._last_move = []
        for direction in selection_path:
            atomic_move = AtomicMove(direction, False)
            atomic_move.is_pusher_selection = True
            self._last_move.append(atomic_move)

        self._selected_pusher = pusher_id

    def move(self, direction):
        """Moves currently selected pusher in ``direction``.

        In `.SolvingMode.FORWARD` mode, pushes the box in front of pusher (if there
        is one).

        In `.SolvingMode.REVERSE` mode pulls box together with pusher (if there is
        one and if ``self.pulls_boxes is True``).

        Args:
            direction (Direction): direction of movement

        Raises:
            .IllegalMoveError: for illegal moves
        """
        options = MoveWorkerOptions()
        if self._solving_mode == SolvingMode.FORWARD:
            options.decrease_pull_count = False
            self._push_or_move(direction, options)
        else:
            options.force_pulls = self._pulls_boxes
            options.increase_pull_count = True
            self._pull_or_move(direction, options)

    def jump(self, new_position):
        """Currently selected pusher jumps to ``new_position``.

        Fails if

            - :class:`.Mover` is in `.SolvingMode.FORWARD` mode
            - pusher can't be dropped on ``new_position``
            - first pull had been made

        Raises:
            .IllegalMoveError: for illegal jumps
        """
        if self._pull_count != 0:
            raise IllegalMoveError("Jumps not allowed after first pull")

        if self._solving_mode != SolvingMode.REVERSE:
            raise IllegalMoveError("Jumps allowed only in reverse solving mode")

        old_position = self._manager.pusher_position(self._selected_pusher)
        if old_position == new_position:
            return

        try:
            self._manager.move_pusher_from(old_position, new_position)
        except CellAlreadyOccupiedError as exc:
            raise IllegalMoveError(str(exc))

        path = self._manager.board.positions_path_to_directions_path(
            self._manager.board.find_jump_path(old_position, new_position)
        )

        def jump_am(direction):
            atomic_move = AtomicMove(direction, False)
            atomic_move.is_jump = True
            atomic_move.pusher_id = self._selected_pusher
            return atomic_move

        self._last_move = [jump_am(direction) for direction in path]

    def undo_last_move(self):
        """
        Takes sequence of moves stored in self.last_move and undoes it.

        See Also:
            `.Mover.last_move`
        """
        new_last_moves = []
        old_last_moves = self._last_move

        jump_key = 0
        pusher_change_key = 1
        move_key = 2

        def key_functor(elem):
            if elem.is_jump:
                return jump_key
            if elem.is_pusher_selection:
                return pusher_change_key
            return move_key

        for moves_type, moves_group in groupby(reversed(old_last_moves), key_functor):
            if moves_type == move_key:
                for atomic_move in moves_group:
                    self._undo_atomic_move(atomic_move)
                    new_last_moves += self._last_move
            elif moves_type == jump_key:
                self._undo_jump(moves_group)
                new_last_moves += self._last_move
            else:
                self._undo_pusher_selection(moves_group)
                new_last_moves += self._last_move

        self._last_move = new_last_moves

    def _undo_atomic_move(self, atomic_move):
        options = MoveWorkerOptions()
        if self._solving_mode == SolvingMode.FORWARD:
            has_box_behind_pusher = self.board_manager.has_box_on(
                self._manager.board.neighbor(
                    self.board_manager.pusher_position(self.selected_pusher),
                    atomic_move.direction,
                )
            )

            if not atomic_move.is_move and not has_box_behind_pusher:
                raise IllegalMoveError("Requested push undo, but no box behind pusher!")
            options.force_pulls = not atomic_move.is_move
            options.increase_pull_count = False
            self._pull_or_move(atomic_move.direction.opposite, options)
        else:
            options.decrease_pull_count = True
            self._push_or_move(atomic_move.direction.opposite, options)

    def _undo_jump(self, jump_moves):
        path = [atomic_move.direction.opposite for atomic_move in jump_moves]
        old_position = self._manager.pusher_position(self._selected_pusher)
        new_position = self._manager.board.path_destination(old_position, path)
        self.jump(new_position)

    def _undo_pusher_selection(self, selection_moves):
        path = [atomic_move.direction.opposite for atomic_move in selection_moves]
        old_position = self._manager.pusher_position(self._selected_pusher)
        new_position = self._manager.board.path_destination(old_position, path)
        self.select_pusher(self._manager.pusher_id_on(new_position))

    def _push_or_move(self, direction, options):
        """
        Perform movement of currently selected pusher in ``direction``.

        In case there is a box in front of pusher, pushes it.
        """
        initial_pusher_position = self._manager.pusher_position(self._selected_pusher)
        in_front_of_pusher = self._manager.board.neighbor(
            initial_pusher_position, direction
        )

        if not in_front_of_pusher:
            raise IllegalMoveError(
                "Can't move pusher off board! (ID: "
                + "{0}, direction: {1})".format(self._selected_pusher, str(direction))
            )

        is_push = False
        in_front_of_box = None
        if self._manager.has_box_on(in_front_of_pusher):
            is_push = True
            in_front_of_box = self._manager.board.neighbor(
                in_front_of_pusher, direction
            )
            if not in_front_of_box:
                raise IllegalMoveError(
                    "Can't push box off board! (ID: "
                    + "{0}, direction: {1})".format(
                        self._manager.box_id_on(in_front_of_pusher), str(direction)
                    )
                )

            try:
                self._manager.move_box_from(in_front_of_pusher, in_front_of_box)
            except CellAlreadyOccupiedError as exc:
                raise IllegalMoveError(str(exc))

        try:
            self._manager.move_pusher_from(initial_pusher_position, in_front_of_pusher)
        except CellAlreadyOccupiedError as exc:
            raise IllegalMoveError(str(exc))

        atomic_move = AtomicMove(direction, is_push)
        atomic_move.pusher_id = self._selected_pusher
        if is_push:
            atomic_move.moved_box_id = self._manager.box_id_on(in_front_of_box)
            if options.decrease_pull_count and self._pull_count > 0:
                self._pull_count -= 1
        self._last_move = [atomic_move]

    def _pull_or_move(self, direction, options):
        """
        Perform movement of currently selected pusher in ``direction``.

        In case there is a box in behind of pusher, might pull it.
        """
        initial_pusher_position = self._manager.pusher_position(self._selected_pusher)
        in_front_of_pusher = self._manager.board.neighbor(
            initial_pusher_position, direction
        )

        if not in_front_of_pusher:
            raise IllegalMoveError(
                "Can't move pusher off board! (ID: "
                + "{0}, direction: {1})".format(self._selected_pusher, str(direction))
            )

        try:
            self._manager.move_pusher_from(initial_pusher_position, in_front_of_pusher)
        except CellAlreadyOccupiedError as exc:
            raise IllegalMoveError(str(exc))

        is_pull = False
        if options.force_pulls:
            behind_pusher = self._manager.board.neighbor(
                initial_pusher_position, direction.opposite
            )
            if behind_pusher and self._manager.board[behind_pusher].has_box:
                is_pull = True
                try:
                    self._manager.move_box_from(behind_pusher, initial_pusher_position)
                except CellAlreadyOccupiedError as exc:
                    raise IllegalMoveError(str(exc))
                if options.increase_pull_count:
                    self._pull_count += 1

        atomic_move = AtomicMove(direction, is_pull)
        atomic_move.pusher_id = self._selected_pusher
        if is_pull:
            atomic_move.moved_box_id = self._manager.box_id_on(initial_pusher_position)
        self._last_move = [atomic_move]
